source/slang/slang-ir-hlsl-legalize.cpp

```diff
@@ -0,0 +1,93 @@
+// slang-ir-hlsl-legalize.cpp
+#include "slang-ir-hlsl-legalize.h"
+
+#include <functional>
+
+#include "slang-ir.h"
+#include "slang-ir-insts.h"
+#include "slang-ir-inst-pass-base.h"
+#include "slang-ir-specialize-function-call.h"
+#include "slang-ir-util.h"
+
+namespace Slang
+{
+
+void searchChildrenForForceVarIntoStructTemporarily(IRModule* module, IRInst* inst)
+{
+    for(auto child : inst->getChildren())
+    {
+        switch(child->getOp())
+        {
+        case kIROp_Block:
+        {
+            searchChildrenForForceVarIntoStructTemporarily(module, child);
+            break;
+        }
+        case kIROp_Call:
+        {
+            auto call = as<IRCall>(child);
+            for(UInt i = 0; i < call->getArgCount(); i++)
+            {
+                auto arg = call->getArg(i);
+                if(arg->getOp() != kIROp_ForceVarIntoStructTemporarily)
+                    continue;
+                auto forceStructArg = arg->getOperand(0);
+                auto forceStructBaseType = as<IRType>(forceStructArg->getDataType()->getOperand(0));
+                if(forceStructBaseType->getOp() == kIROp_StructType)
+                {
+                    arg->replaceUsesWith(arg->getOperand(0));
+                    arg->removeAndDeallocate();
+                    continue;
+                }
+
+                // Replace all non struct inputs with following logic
+                // {`Non-struct var` to `struct var` copy }
+                // {Use `struct var` made as parameter to kIROp_Call}
+                // if(parameterIsPointerLikeTypeWhichMayHaveSideEffectsOnParameter) {`Struct var` to `non-struct` var copy}
```

Comment:
Can you use more plain English?
I don't know if I understood correctly or not but it could be something like this,
```
When the downstream function expects a struct-type and a given parameter is not a struct-type, we create a temporary struct variable.
If the parameter has a "out" modifier, the result needs to be copied back to the parameter.
```